### ISSUE TITLE
Significant performance reduction

### ISSUE THREAD

**Issue by kshamko** [containous/traefik#3422]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

Want to use traefik as LB for high load app.
Tried to reproduce benchmarks  from https://docs.traefik.io/benchmarks/ and discovered that version 1.6 gives about 3000 rps on my desktop. Version 1.5 gives the same result. But on 1.4 I've got 20000 rps. 

#### wrk output (v1.6)
```console
$ wrk -t20 -c1000 -d60s -H "Host: localhost" --latency  http://localhost:8001/bench
Running 1m test @ http://localhost:8001/bench
  20 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   316.46ms  270.26ms   1.99s    61.62%
    Req/Sec   176.00     73.64     1.54k    71.34%
  Latency Distribution
     50%  299.91ms
     75%  484.84ms
     90%  685.63ms
     99%    1.06s 
  206835 requests in 1.00m, 19.92MB read
  Socket errors: connect 0, read 0, write 0, timeout 22
Requests/sec:   3442.57
Transfer/sec:    339.55KB
```
During the test CPU usage was about 50% for each core

#### wrk output (v1.4)
```console
$ wrk -t20 -c1000 -d60s -H "Host: localhost" --latency  http://localhost:8001/bench
Running 1m test @ http://localhost:8001/bench
  20 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    61.84ms   97.45ms   1.44s    95.72%
    Req/Sec     1.08k   352.23     5.61k    86.44%
  Latency Distribution
     50%   43.17ms
     75%   58.30ms
     90%   79.00ms
     99%  596.76ms
  1277025 requests in 1.00m, 123.01MB read
  Non-2xx or 3xx responses: 85
Requests/sec:  21252.91
Transfer/sec:      2.05MB
```
During the test CPU usage was about 95% for each core

### What did you expect to see?

Expected RPS value to be about 20k

### What did you see instead?

RPS value 3k

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      v1.6.2
Codename:     tetedemoine
Go version:   go1.10.2
Built:        2018-05-22_03:19:06PM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

traefik.toml

```toml
maxIdleConnsPerHost = 100000
defaultEntryPoints = ["http"]
logLevel = "DEBUG"

[entryPoints]
  [entryPoints.http]
  address = ":8001"

[file]
[backends]
  [backends.backend1]
    [backends.backend1.servers.server1]
    url = "http://whoami:80"
    weight = 1
    [backends.backend1.servers.server2]
    url = "http://whoami:80"
    weight = 1

[frontends]
  [frontends.frontend1]
  backend = "backend1"
    [frontends.frontend1.routes.test_1]
    rule = "Host: localhost"
```

docker-compose.yml
```yml
version: '2'

services:
  #The reverse proxy service (Træfik)
  reverse-proxy:
    image: traefik:1.4 #The official Traefik docker image
    command: --debug -c /etc/traefik/traefik.toml #Enables the web UI and tells Træfik to listen to docker
    ports:
      - "80:80"     #The HTTP port
      - "8080:8080" #The Web UI (enabled by --api)
      - "8001:8001"
    volumes:
      #- /var/run/docker.sock:/var/run/docker.sock #So that Traefik can listen to the Docker events
      - ./traefik.toml:/etc/traefik/traefik.toml
  
  #A container that exposes a simple API
  whoami:
    image: emilevauge/whoami #A container that exposes an API to show it's IP address
    labels:
      - "traefik.frontend.rule=Host:whoami.docker.localhost"
```

### If applicable, please paste the log output in DEBUG level (`--logLevel=DEBUG` switch)

```shell
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=info msg="Using TOML configuration file /etc/traefik/traefik.toml" 
whoami_1         | Starting up on port 80
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=info msg="Traefik version v1.4.6 built on 2018-01-02_12:27:08PM" 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=debug msg="Global configuration loaded {"GraceTimeOut":10000000000,"Debug":true,"CheckNewVersion":true,"AccessLogsFile":"","AccessLog":null,"TraefikLogsFile":"","LogLevel":"DEBUG","EntryPoints":{"http":{"Network":"","Address":":8001","TLS":null,"Redirect":null,"Auth":null,"WhitelistSourceRange":null,"Compress":false,"ProxyProtocol":null,"ForwardedHeaders":{"Insecure":true,"TrustedIPs":null}}},"Cluster":null,"Constraints":[],"ACME":null,"DefaultEntryPoints":["http"],"ProvidersThrottleDuration":2000000000,"MaxIdleConnsPerHost":100000,"IdleTimeout":0,"InsecureSkipVerify":false,"RootCAs":null,"Retry":null,"HealthCheck":{"Interval":30000000000},"RespondingTimeouts":null,"ForwardingTimeouts":null,"Docker":null,"File":{"Watch":true,"Filename":"/etc/traefik/traefik.toml","Constraints":null,"Trace":false,"DebugLogGeneratedTemplate":false,"Directory":""},"Web":null,"Marathon":null,"Consul":null,"ConsulCatalog":null,"Etcd":null,"Zookeeper":null,"Boltdb":null,"Kubernetes":null,"Mesos":null,"Eureka":null,"ECS":null,"Rancher":null,"DynamoDB":null}" 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=info msg="Preparing server http &{Network: Address::8001 TLS:<nil> Redirect:<nil> Auth:<nil> WhitelistSourceRange:[] Compress:false ProxyProtocol:<nil> ForwardedHeaders:0xc4202d8660} with readTimeout=0s writeTimeout=0s idleTimeout=3m0s" 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=info msg="Starting provider *file.Provider {"Watch":true,"Filename":"/etc/traefik/traefik.toml","Constraints":null,"Trace":false,"DebugLogGeneratedTemplate":false,"Directory":""}" 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=info msg="Starting server on :8001" 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=debug msg="Validation of load balancer method for backend backend1 failed: invalid load-balancing method ''. Using default method wrr." 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=debug msg="Configuration received from provider file: {"backends":{"backend1":{"servers":{"server1":{"url":"http://whoami:80","weight":1},"server2":{"url":"http://whoami:80","weight":1}},"loadBalancer":{"method":"wrr"}}},"frontends":{"frontend1":{"entryPoints":["http"],"backend":"backend1","routes":{"test_1":{"rule":"Host: localhost"}},"priority":0,"basicAuth":null,"headers":{}}}}" 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=debug msg="Last file config received more than 2s, OK" 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=debug msg="Creating frontend frontend1" 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=debug msg="Wiring frontend frontend1 to entryPoint http" 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=debug msg="Creating route test_1 Host: localhost" 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=debug msg="Creating backend backend1" 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=debug msg="Creating load-balancer wrr" 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=debug msg="Creating server server1 at http://whoami:80 with weight 1" 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=debug msg="Creating server server2 at http://whoami:80 with weight 1" 
reverse-proxy_1  | time="2018-06-01T08:49:37Z" level=info msg="Server configuration reloaded on :8001" 
```

**Comment by mmatur** [comment]
Hi @kshamko,

First of all, thanks for your in interest in Træfik.

In your configuration, I see that you have enabled `--debug` and that you have set `logLevel = "DEBUG"`. For the sake of the benchmark, my recommendation is to avoid this.

I also see that you want to benchmark Træfik with Traefik and whoamI running in docker. It is possible to do so, but be aware that the docker networking has an impact on the benchmark results.
You also use the DNS name of the whoamI service instead of its IP address, which also has an impact (the DNS resolution takes some time).

Between 1.4 and 1.6, we have made (a lot) more information available in the “debug” logs, and since you have enabled them, it explains why you see such drastic changes.

For the sake of the example, in Traefik 1.4 the round trip debug log looked like:
```console
reverse-proxy_1  | time="2018-06-01T11:57:40Z" level=debug msg="Round trip: http://whoami:80, code: 200, duration: 2.836759ms"
```

In Træfik version 1.6, the same debug log now looks like:
```console
reverse-proxy_1  | time="2018-06-01T11:58:25Z" level=debug msg="vulcand/oxy/roundrobin/rr: competed ServeHttp on request" Request="{\"Method\":\"GET\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/bench\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\"},\"Proto\":\"HTTP/1.1\",\"ProtoMajor\":1,\"ProtoMinor\":1,\"Header\":{},\"ContentLength\":0,\"TransferEncoding\":null,\"Host\":\"localhost\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"172.20.0.1:58896\",\"RequestURI\":\"/bench\",\"TLS\":null}"
```

So naturally, more information means that it takes Traefik more time to write the data in the stdout.
I did some test on my computer, and if you’d like, you can test this out for yourself:

## docker-compose.yml
```yaml
version: '2'

services:
  traefik:
    image: traefik:1.6
    ports:
      - "80:80"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ./traefik.toml:/etc/traefik/traefik.toml
  whoami:
    image: emilevauge/whoami
    labels:
      - "traefik.enable=true"
      - "traefik.frontend.rule=Host:localhost"
      - "traefik.port=80"
```

## traefik.toml
```toml
maxIdleConnsPerHost = 100000
defaultEntryPoints = ["http"]

[entryPoints]
  [entryPoints.http]
  address = ":80"

[docker]
exposedByDefault = false
```

# Results on my laptop :

## Træfik v1.4
```console
wrk -t20 -c1000 -d60s -H "Host: localhost" --latency  http://127.0.0.1:80/bench  
Running 1m test @ http://127.0.0.1:80/bench
  20 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    30.64ms   34.11ms   1.34s    97.48%
    Req/Sec     1.72k   245.08     5.22k    85.04%
  Latency Distribution
     50%   28.17ms
     75%   39.06ms
     90%   50.57ms
     99%   75.38ms
  2047496 requests in 1.00m, 197.22MB read
  Socket errors: connect 0, read 0, write 0, timeout 24
Requests/sec:  34072.75
Transfer/sec:      3.28MB
```

## Træfik v1.6
```console
wrk -t20 -c1000 -d60s -H "Host: localhost" --latency  http://127.0.0.1:80/bench
Running 1m test @ http://127.0.0.1:80/bench
  20 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    48.84ms   57.75ms   1.34s    97.10%
    Req/Sec     1.12k   165.51     3.43k    79.09%
  Latency Distribution
     50%   43.57ms
     75%   67.08ms
     90%   82.93ms
     99%  131.85ms
  1337740 requests in 1.00m, 128.85MB read
  Socket errors: connect 0, read 0, write 0, timeout 19
Requests/sec:  22261.41
Transfer/sec:      2.14MB
```

As you can see, there are still differences between the two versions, but this is something we’re aware of (and something that is on the roadmap).

**Comment by Juliens** [comment]
I made a PR in order to handle those performance issue.

Here are my results with docker-compose on my laptop (remember that there is an overhead because of the docker network)

## docker-compose.yml
```yaml
version: '2'

services:
  traefik:
    image: traefik:1.4
    ports:
      - "80:80"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ./traefik.toml:/etc/traefik/traefik.toml
  whoami:
    image: emilevauge/whoami:v1.0.1
    labels:
      - "traefik.enable=true"
      - "traefik.frontend.rule=Host:localhost"
      - "traefik.port=80"
```

## traefik.toml
```toml
maxIdleConnsPerHost = 100000
defaultEntryPoints = ["http"]

[entryPoints]
  [entryPoints.http]
  address = ":80"

[docker]
exposedByDefault = false
```

# Results on my laptop :

## Traefik 1.4

```bash
wrk -t20 -c1000 -d60s -H "Host: localhost" --latency  http://127.0.0.1:80/bench
Running 1m test @ http://127.0.0.1:80/bench
  20 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    38.64ms   32.11ms   1.09s    83.75%
    Req/Sec     1.35k   147.79     3.10k    73.53%
  Latency Distribution
     50%   36.44ms
     75%   53.69ms
     90%   66.28ms
     99%   96.57ms
  1609022 requests in 1.00m, 156.52MB read
Requests/sec:  26778.55
Transfer/sec:      2.60MB 
```

## The PR version

```bash
wrk -t20 -c1000 -d60s -H "Host: localhost" --latency  http://127.0.0.1:80/bench
Running 1m test @ http://127.0.0.1:80/bench
  20 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    34.72ms   37.41ms   1.10s    98.76%
    Req/Sec     1.52k   167.12     7.32k    80.55%
  Latency Distribution
     50%   34.82ms
     75%   44.22ms
     90%   52.07ms
     99%   73.60ms
  1815044 requests in 1.00m, 176.56MB read
Requests/sec:  30207.06
Transfer/sec:      2.94MB
```

**Comment by traefiker** [comment]
Closed by #3450.
